### PR TITLE
[Improve](row-store) support row cache

### PR DESCRIPTION
**Pull request by eldenmoon**
Currently only support PageCache of columns, but for row store we need a seperate cache to increase cache hit rate

# Proposed changes

Issue Number: close #xxx

## Problem summary

Describe your changes.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [ ] No
    - [ ] I don't know
2. Has unit tests been added:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [ ] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [ ] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 34.41 seconds
 load time: 510 seconds
 storage size: 17123073238 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230130162356_clickbench_pr_87287.html

File: be/src/common/config.h [hunk 1]
```diff
@@ -257,6 +258,8 @@ CONF_Int32(storage_page_cache_shard_size, "16");
 CONF_Int32(index_page_cache_percentage, "10");
 // whether to disable page cache feature in storage
 CONF_Bool(disable_storage_page_cache, "false");
+// whether to disable row cache feature in storage
+CONF_Bool(disable_storage_row_cache, "false");
```
Comment (cambyzju) [inline be/src/common/config.h:262]: If row cache do not used in most cases, please disable it by default.

Comment (eldenmoon) [reply be/src/common/config.h:262]: row cache is used for row store, row store  is disabled by default, so for most cases this flag is unused

File: be/src/vec/jsonb/serialize.cpp [hunk 2]
```diff
@@ -319,4 +319,22 @@ void JsonbSerializeUtil::jsonb_to_block(const TupleDescriptor& desc,
     }
 }
 
+// single row
+void JsonbSerializeUtil::jsonb_to_block(const TupleDescriptor& desc, const Slice& data,
```
Comment (xiaokang) [inline be/src/vec/jsonb/serialize.cpp:323]: jsonb_to_block for single row can be reused by jsonb_to_block for multiple rows

File: be/src/common/config.h [hunk 3]
```diff
@@ -246,6 +246,7 @@ CONF_mBool(row_nums_check, "true");
 // modify them upon necessity
 CONF_Int32(min_file_descriptor_number, "60000");
 CONF_Int64(index_stream_cache_capacity, "10737418240");
+CONF_String(row_cache_mem_limit, "20%");
```
[the remark this answers is not in the document]
Comment (eldenmoon) [reply be/src/common/config.h:249]: high concurrent point query need memory to increase cache hit rate, I think 20% is OK for such work load

File: be/src/vec/jsonb/serialize.cpp (hunk 2 above)
Comment (eldenmoon) [reply be/src/vec/jsonb/serialize.cpp:323]: ok

File: be/src/olap/rowset/segment_v2/segment_writer.cpp [hunk 4]
```diff
@@ -252,8 +253,13 @@ Status SegmentWriter::append_block(const vectorized::Block* block, size_t row_po
...
             // create primary indexes
             for (size_t pos = 0; pos < num_rows; pos++) {
-                RETURN_IF_ERROR(
-                        _primary_key_index_builder->add_item(_full_encode_keys(key_columns, pos)));
+                const std::string& key = _full_encode_keys(key_columns, pos);
+                RETURN_IF_ERROR(_primary_key_index_builder->add_item(key));
+                if (!config::disable_storage_row_cache && _tablet_schema->store_row_column() &&
+                    _opts.is_direct_write) {
+                    // invalidate cache
+                    RowCache::instance()->erase({_opts.rowset_ctx->tablet_id, key});
```
[the remark this answers is not in the document]
Comment (eldenmoon) [reply be/src/olap/rowset/segment_v2/segment_writer.cpp:261]: can not insert value because  load may be failed, and rowset is not visible

**Review by xiaokang — APPROVED** [review]
LGTM

**Review by Gabriel39 — APPROVED** [review]
LGTM

**Pull request merged**